### DISCUSSION TITLE
CSNN loss function for image input to image output [2D -> 2D]

### DISCUSSION THREAD

**Discussion by Radu2k** [jeshraghian/snntorch#145] [Q&A] [answered]
Hello,

I am working on an implementation of a CSNN network using rate encoding but the output is an image, so the output is 
[num_steps, batch_size, out_size1, out_size2].

I was looking around for a loss function but the only one that worked for the first two iterations was Cosine Similarity. As problems appear when backpropagating due to the fact that the output is unsqueezed to compress my rate encoding I was wandering if any of the SF have this sort of functionality?

Thanks!

**Comment by jeshraghian** [comment, marked as answer]
Hi Radu,

We haven't natively implemented any losses yet into SF that account for
problems outside of classification, though it's definitely an important
feature.

I suppose the loss would need to manually constructed. Eg, sum up the
spikes in the output, divide it by the number of time steps to normalise
everything to 1, and then apply cosine similarity or mean square error to
compare it to the original image.

There is some recent work on SNN autoencoders, so I'll look into those to
see how SF can be expanded to handle losses acting on multidimensional
outputs.

**Reply by Radu2k** [reply]
Hi Jason,

That sounds good to me. Looking forward to seeing this implemented. If there is any way to help implementing this please let me know.
